rpc: retain information about failed connections

tbg opened the pull request:
Prior to this change, `rpc.Context` did not remember a connection after it
failed. A connection attempt would be created whenever we didn't have a
functioning connection to the given `(nodeID, targetAddress, class)` and
callers would be multiplexed onto it as long as that attempt was not known to
have failed.

The main problem with this setup was that it makes it difficult to provide good
observability because a remote node that just hasn't been dialed yet and one
that is down but between attempts looks the same. We could write code that
periodically tries to fully connect all nodes in the cluster to each other, but
even then exporting good metrics is challenging and while we're currently
comfortable with a densely connected cluster, that may change over time.

An additional challenge was presented by circuit breaking and logging of
attempts. Without state retained, we were limited to a simple retry scheme with
lots of redundant logging. It wasn't easy to log how long a connection had been
unhealthy (as an unhealthy connection was represented as an absent connection),
so in effect folks had to trawl through logs to grab the corresponding
timestamps of first and last failure.

Another piece of complexity were the RPC circuit breakers. These were
implemented at the NodeDialer-level (i.e. one circuit breaker per
`(NodeID,Class`)) but kept in the `rpc.Context` (which generally operates on
`(NodeID, Addr, Class)` because gossip also used them. The library they were
using uses recruitment of client traffic, which also isn't ideal as it could
periodically subject a SQL query to a timeout failure. We'd really prefer it
if the breakers probed without user traffic recruitment.

This PR addresses the above shortcomings:

- state is now retained across restarts on a `(NodeID, Addr, Class)` basis in
  the `rpc.Context. This has a few complications, for example we need to handle
decommissioned nodes, as well as nodes that restart under a new listening
address.
- the NodeDialer-level circuit breakers are removed.
- we're no longer recruiting user traffic for probes. Instead, we adopt the
  `util/circuit` package already used for Replica-level circuit breaking. This
library uses an async background probe to determine when to heal the breaker.

I also think the amount of incidental complexity in the `rpc` package is greatly
reduced as a result of this work.

I ran the `failover/non-system/` suite of tests to validate this change. The results
look good[^res].

[^res]: https://gist.github.com/tbg/5f5c716fa2fe5a21eca4a0c3a4a35069

Epic: CRDB-21710
Release note (ui change): a "Networking" metrics dashboard was added.
Release note (ops change): CockroachDB now exposes the following metrics:
- `rpc.connection.{un,}healthy`: Gauge of healthy/unhealthy connections,
- `rpc.connection.{un,}healthy_nanos`: Gauge of aggregate duration the currently {un,}healthy connections have been {,un}healthy for,
-  `grpc.connection.avg_round_trip_latency`: sum of weighted moving averages of round-trip latencies for all peers (it really only makes sense to consume these on a per-peer basis via prometheus)
- Counters `rpc.connection.heartbeats` and `rpc.connection.failures`.
When the `server.child_metrics.enabled` cluster setting is set, these metrics export per-peer values, meaning they can be used to derive a connectivity graph.
Release note (ops change): logging for intra-node RPC connection failures has improved by reducing frequently and improving information content.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/99191)
<!-- Reviewable:end -->

tbg committed:
kvpb: introduce NodeDecommissionedError

This will be used in a follow-up commit to more precisely communicate that a
connection attempt failed due to a decommissioned peer or source.

It will be more important to get this right as we switch to a stateful
connection pool. We don't want to attempt to reconnect to decommissioned nodes
forever, and we also don't want to misrepresent them as "unavailable" for the
purpose of network observability.

So far we're relying on the `PermissionDenied` error code but this is already
overloaded. It's better to provide an explicit gRPC status payload that leaves
no room for interpretation.

Epic: none
Release note: None

tbg committed:
rpc: retain information about failed connections

Prior to this change, `rpc.Context` did not remember a connection after it
failed. A connection attempt would be created whenever we didn't have a
functioning connection to the given `(nodeID, targetAddress, class)` and
callers would be multiplexed onto it as long as that attempt was not known to
have failed.

The main problem with this setup was that it makes it difficult to provide good
observability because a remote node that just hasn't been dialed yet and one
that is down but between attempts looks the same. We could write code that
periodically tries to fully connect all nodes in the cluster to each other, but
even then exporting good metrics is challenging and while we're currently
comfortable with a densely connected cluster, that may change over time.

An additional challenge was presented by circuit breaking and logging of
attempts. Without state retained, we were limited to a simple retry scheme with
lots of redundant logging. It wasn't easy to log how long a connection had been
unhealthy (as an unhealthy connection was represented as an absent connection),
so in effect folks had to trawl through logs to grab the corresponding
timestamps of first and last failure.

Another piece of complexity were the RPC circuit breakers. These were
implemented at the NodeDialer-level (i.e. one circuit breaker per
`(NodeID,Class`)) but kept in the `rpc.Context` (which generally operates on
`(NodeID, Addr, Class)` because gossip also used them. The library they were
using uses recruitment of client traffic, which also isn't ideal as it could
periodically subject a SQL query to a timeout failure. We'd really prefer it
if the breakers probed without user traffic recruitment.

This PR addresses the above shortcomings:

- state is now retained across restarts on a `(NodeID, Addr, Class)` basis in
  the `rpc.Context. This has a few complications, for example we need to handle
decommissioned nodes, as well as nodes that restart under a new listening
address.
- the NodeDialer-level circuit breakers are removed.
- we're no longer recruiting user traffic for probes. Instead, we adopt the
  `util/circuit` package already used for Replica-level circuit breaking. This
library uses an async background probe to determine when to heal the breaker.

Epic: CRDB-21710
Release note: TODO(during review): explain the new metrics and any other changes.

tbg committed:
nodedialer: remove test for deprecated breaker from TestConnHealth{,TryDial}

Epic: none
Release note: None

tbg committed:
nodedialer: modernize TestDialNoBreaker

Test the functionality via the new circuit breakers, i.e. this test will
continue working if we remove the old breakers (which it no longer tests).

Epic: none
Release note: None

tbg committed:
nodedialer: remove TestConcurrentCancellationAndTimeout

It was specific to the nodedialer-level circuit breakers, which will be
removed.

Epic: none
Release note: None

tbg committed:
nodedialer: remove breaker check from TestNodedialerPositive

This was testing the deprecated breaker.

Epic: none
Release note: None

tbg committed:
nodedialer: remove Test{Disconnects,ResolverErrors}Trip

Epic: none
Release note: None

tbg committed:
importer: use new rpc breaker in TestImportRetriesBreakerOpenFailure

Epic: none
Release note: None

tbg committed:
nodedialer: remove use of old breaker in TestConnHealthInternal

Epic: none
Release note: None

tbg committed:
kvserver: use new breakers in RaftTransport

Epic: none
Release note: None

tbg committed:
kvserver: remove outdated comment

The previous breaker was traffic-recruiting, and so this deadlocks

```go
if breaker.Tripped(nodeID) {
  return errors.New("tripped")
}
dial(nodeID)
```

because the dial will never be attempted.

The new breakers don't have this issue because they're running a probe in the
background (which is triggered by any inquiry to the breaker), so the breaker
will untrip once the connection is ready to be used (at which point the dial
will succeed).

Epic: none
Release note: None

tbg committed:
nodedialer: remove uses of deprecated breaker

Epic: none
Release note: None

tbg committed:
gossip: don't use deprecated breaker

Epic: none
Release note: None

tbg committed:
nodedialer,rpc: remove deprecated breaker

Farewell, old friend.

Epic: none
Release note: None

tbg committed:
rpc: update the round-trip-latency histogram

improve documentation.

Epic: None
Release note: None

tbg commented:
Sneak Peek of metrics

<img width="1021" alt="image" src="https://github.com/cockroachdb/cockroach/assets/5076964/2bdcf627-3ab2-4dd2-aed1-700b47458cfe">

Here are the prometheus metrics, with <source,dest> labels (needs `set cluster setting server.child_metrics.enabled = true;`)

https://gist.github.com/tbg/c7ec86c0d180580f17756a51fb5944e4

aliher1911 commented on pkg/rpc/breaker_new_probe.go:
Why do we touch all peers here? How does failure of this connection link us to other potential peers?

tbg replied:
It's just a convenient trigger, added this comment (with next push)

	// There might be other peers in the map that are pending deletion, but which
	// are no longer seeing activity. To eventually clear them out, we check all
	// conns when any conn fails. This avoids the need to have an extra goroutine
	// sitting in all of the goroutine stacks we look at during support.
	touchOldPeers(p.peers, now)

aliher1911 commented on pkg/rpc/context.go:
```diff
@@ -2719,26 +2748,224 @@ func (rpcCtx *Context) loadOrCreateConnAttempt(
+// returned.
+//
+// TODO(during review): move after dust has settled.
+func hasSiblingConn(peers map[peerKey]*peer, self peerKey) (healthy, ok bool) {
```
Needs to be named *Locked to show that we must lock peers prior to calling this as `shouldDeleteAfter` does.

aliher1911 commented on pkg/rpc/context.go:
```diff
@@ -2719,26 +2748,224 @@ func (rpcCtx *Context) loadOrCreateConnAttempt(
+	conns.mu.RLock()
+	defer conns.mu.RUnlock()
+
+	sibHealthy, ok := hasSiblingConn(conns.mu.m, myKey)
```
Is it ok that we could wait for sibling connection to get its health resolved here? Wouldn't that delay resolving gatedCC below.

aliher1911 replied:
Maybe `rpc.Connection.wait` needs a better name then because that got me confused since I didn't look on the implementation. Something like `waitOrDefault`?

tbg replied:
Good idea, done.

aliher1911 reviewed (commented):
Some initial questions on the first pass of new breakers.

tbg commented on pkg/rpc/context.go:
```diff
@@ -2237,124 +2379,56 @@ func (rpcCtx *Context) GRPCDialPod(
+	// Won race. Actually create a peer.
 
-	newConn := newConnectionToNodeID(k.nodeID, k.class)
+	if conns.mu.m == nil {
```
You can read a `nil` map just fine, the call you allude to bottoms out here:

```go
func (m *peerMap) getRLocked(k peerKey) (PeerSnap, bool) {
	p, ok := m.mu.m[k]
	if !ok {
		return PeerSnap{}, false
	}
	return p.snap(), true
}
```

That works fine, even though `m.mu.m` is nil. We basically make the map right as we're about to do the first write to it (which will panic on a nil map).

I can also "always" make the map in `NewContext`, maybe that's easier. The pattern I'm using here is more useful when the map remains unused most of the time, but here it's definitely going to be used. Your call.

tbg commented on pkg/rpc/context.go:
```diff
@@ -2214,248 +2375,197 @@ func (rpcCtx *Context) GRPCDialPod(
+	// again (but keep running). This means we are immediately able to hand out
+	// the connection to callers, who will then block on the initialHeartbeatDone
+	// channel (as opposed to bouncing off the breaker).
+	_ = (*breakerProbe)(nil).launch // [1]
```
Yeah just for ease of navigation. Let me know if you have a comment in mind that would help (one that isn't too long!) or whether you think this is not necessary.

tbg replied:
Removed it.

aliher1911 commented on pkg/rpc/context_test.go:
```diff
@@ -2655,47 +2514,147 @@ func TestHeartbeatDialback(t *testing.T) {
+		return nil
+	}
+
+	if hf := m.ConnectionHealthyFor.Value(); (healthy > 0) != (hf != 0) {
```
`(healthy > 0) != (hf != 0)` is not obvious at all. Not sure we can simplify the condition, but at least a comment saying that `checkDurations` mean HealthyFor, UnhealthyFor and RoundTrip must be non zero if corresponding counters are non zero.

tbg replied:
I added this comment:

	// HealthyFor is nonzero if and only if there is at least one healthy connection, since it's
	// the sum of the connections' HealthyFor, and on a connection this is nonzero if and only
	// if the connection is healthy.

Does that help or do you think further clarification is necessary? The condition is exact, so all we can do is rewrite it. `hf > 0` is probably better than `hf != 0`, I made that change.

aliher1911 reviewed (commented):
I like how it is now looking overall, but the fact that we have so many "connection" entities is odd. It is not different from existing design though. What I think odd is that we have context that gives us connection on dial but then you need to connect to get actual connection. In socket API's you can get unconnected/unbound socket that you can connect, but you still use the same object. In our case we get potentially disconnected connection from context/dialer and then we connect and get grpc connection which is bound to this connection object.
It looks more like a factory of some sort as it doesn't provide any functionality that connection imply i.e. sending data.

Comments on `func (n *Dialer) dial(` seem outdated as its lost ins built in breaker functionality, but still mentions nil where argument was replaced by a bool. Maybe it should just describe things on higher level without going into breaker detail?

aliher1911 commented:
Mentioned briefly offline, but worth spelling here for posterity.

I think using cirtuit.Breaker to drive probe lifecycle is odd. It adds unnecessary complexity with callbacks. I think it would be easier to just start a probe when connection is established and let it inform breaker which will be used without its asyncStartProbe part. We can still get standardized logging events and state tracking out of it.

tbg committed:
circuit: allow triggering probe

This will be used in the `rpc` package to remove an ugly workaround.

Epic: None
Release note: None

tbg commented on pkg/rpc/context.go:
```diff
@@ -302,71 +311,184 @@ func NewServerEx(
 	}, nil
 }
 
+type gatedClientConn struct {
```
Yes, it's just a future. I renamed it to `connFuture`, that's a better name, thanks for the hint.

tbg commented on pkg/rpc/testdata/TestReconnection/initial.txt:
```diff
@@ -0,0 +1,59 @@
+----
+ok
+
+connect class=def
```
I added some comments, not sure this is what you had in mind, PTAL.

tbg commented on pkg/rpc/datadriven_test.go:
```diff
@@ -0,0 +1,230 @@
+			env.clock.Advance(time.Second)
+			switch d.Cmd {
+			case "dial":
+				env.handleDial(scanClass(t, d))
```
I am about to push a significant update to the datadriven test, see main thread.

tbg commented:
Pushed another update - improving the datadriven test and fixing bugs/adding logging that came out of that. I feel pretty good about all of this working now.

The first new commit is:

> 33e89b29f7e TODO(tbg): review the preceding commits, I think they all amend into the wrong one.

tbg committed:
rpc: remove TestReconnectAfterAddressChange

There is now stronger coverage in the datadriven tests, see `TestReconnection`.

Epic: none
Release note: None

aliher1911 commented on pkg/rpc/context_test.go:
```diff
@@ -788,453 +800,51 @@ func BenchmarkInternalClientAdapter(b *testing.B) {
-
-// TestHeartbeatHealth verifies that the health status changes after
-// heartbeats succeed or fail due to transport failures.
-func TestHeartbeatHealthTransport(t *testing.T) {
```
This test is still referenced in comments in this file grpcRunKeepaliveTestCase()

aliher1911 commented on pkg/rpc/datadriven_test.go:
```diff
@@ -0,0 +1,342 @@
+				env.handleShow(ctx, env.lookupTarget(t, d.CmdArgs...), scanClass(t, d))
+			case "set-hb-err":
+				var dc bool
+				if d.HasArg("decommissioned") {
```
```
d.MaybeScanArgs(t, "decommissioned", &dc)
```

aliher1911 commented on pkg/rpc/datadriven_test.go:
```diff
@@ -0,0 +1,342 @@
+				var healthy int64
+				var unhealthy int64
+				var inactive int64
+				if d.HasArg("healthy") {
```
MaybeScanArgs as well

aliher1911 commented on pkg/rpc/datadriven_test.go:
```diff
@@ -0,0 +1,230 @@
+	grpcstatus "google.golang.org/grpc/status"
+)
+
+func TestReconnection(t *testing.T) {
```
Is it out of date already? 🤣  Maybe mention n1'' convention? Looks neat in test data but magical until you see that we have preset 2 servers with same name that you skip.

tbg replied:
Fixed in next push :)

aliher1911 reviewed (commented):
Some comments about tests otherwise looks great!

aliher1911 commented on pkg/rpc/peer_test.go:
```diff
@@ -0,0 +1,204 @@
+		// Peer is not sibling of other peers that have non-matching target addrs.
+		{
+			name: "zero-unrelated",
+			m: (&pmb{}).
```
Fluent map builder is fancy, but isn't `pmb(defSelf(), p(n,t,c,h))` as a func with varargs more readable because less () and {}?
I like simple, but YMMV so up to you.

tbg replied:
You're right, but it's a little awkward because the mock `peer` I construct here doesn't know its own `connKey`. I still think your suggestion is directionally right, but I'm going to leave as is.

aliher1911 reviewed (approved):
🎉🎉🎉 LGTM 🎉🎉🎉 
🍿

tbg commented:
bors r=aliher1911

tbg commented:
Thanks I'll take a look tomorrow!

tbg commented:
This should hopefully fix it once merged: https://github.com/cockroachdb/cockroach/pull/104945

data-matt commented:
Folks, any chance of backporting this flag to 22.2? thanks

tbg commented:
No, this is a major change. We won't be able to backport it.

data-matt commented:
@tbg , the cluster setting `server.child_metrics.enabled` will be in 23.1+ ? I see references to 23.1 here.

Thanks cc @florence-crl

erikgrinaker commented:
>@tbg , the cluster setting server.child_metrics.enabled will be in 23.1+ ? I see references to 23.1 here.

The setting `server.child_metrics.enabled` has existed for years, and will enable additional Prometheus metrics for possibly-high-cardinality metrics. For example, changefeeds appear to use them to expose additional per-changefeed metrics.

However, prior to 23.2, this setting does not expose additional RPC metrics. In 23.2, it will also expose per-peer connection metrics as described in the release notes in this PR. This does not happen in 23.1, and can not since the necessary changes are too invasive.

The 23.1 reference you see here was for a minor bugfix that was extracted out of this PR, see https://github.com/cockroachdb/cockroach/pull/105626. It is not related to the functionality you're interested in.

data-matt commented:
Thanks for the clear answer @erikgrinaker.
